src/coreclr/vm/jitinterface.cpp

```diff
@@ -1644,49 +1654,82 @@ static uint32_t ThreadLocalOffset(void* p)
     uint8_t* pOurTls = pTls[_tls_index];
     return (uint32_t)((uint8_t*)p - pOurTls);
 }
+#elif TARGET_OSX
+extern "C" void* GetThreadVarsSectionOffset();
 
-void CEEInfo::getThreadLocalStaticBlocksInfo (CORINFO_THREAD_STATIC_BLOCKS_INFO* pInfo, bool isGCType)
+void* getThreadVarsSectionAddress(uint8_t* p)
 {
-    CONTRACTL {
-        NOTHROW;
-        GC_NOTRIGGER;
-        MODE_PREEMPTIVE;
-    } CONTRACTL_END;
+    if (!(p[0] == 0x48 && p[1] == 0x8d && p[2] == 0x3d))
+    {
+        // The optimization is disabled if coreclr is not compiled in .so format.
+        return 0;
+    }
 
-    JIT_TO_EE_TRANSITION_LEAF();
+    // At this point, `p` contains the instruction pointer and is pointing to the above opcodes.
+    // These opcodes are patched by the dynamic linker.
+    // Move beyond the opcodes that we have already checked above.
+    p += 3;
 
-    pInfo->tlsIndex.addr = (void*)static_cast<uintptr_t>(_tls_index);
-    pInfo->tlsIndex.accessType = IAT_VALUE;
+    // The descriptor address is located at *p at this point.
+    // (p + 4) below skips the descriptor address bytes embedded in the instruction and
+    // add it to the `instruction pointer` to find out the address.
+    return *(uint32_t*)p + (p + 4);
+}
 
-    pInfo->offsetOfThreadLocalStoragePointer = offsetof(_TEB, ThreadLocalStoragePointer);
-    if (isGCType)
-    {
-        pInfo->offsetOfThreadStaticBlocks = ThreadLocalOffset(&t_GCThreadStaticBlocks);
-        pInfo->offsetOfMaxThreadStaticBlocks = ThreadLocalOffset(&t_GCMaxThreadStaticBlocks);
-    }
-    else
+void* getThreadVarsSectionOffset()
+{
+#ifdef TARGET_ARM64
+    return reinterpret_cast<void*>(GetThreadVarsSectionOffset());
+#else
+    // On x64, the address is related to rip, so, disassemble the function,
```

Comment:
Can the asm helper be change to do `lea rdi, _t_ThreadStatics@TLVP[rip]` to avoid the disassembling on OSX? Or does the linker choke it on `lea`?

Reply:
> Can the asm helper be change to do `lea rdi, _t_ThreadStatics@TLVP[rip]` to avoid the disassembling on OSX?

We still end up with the following and I could change the logic to instead read the address of `GetThreadVarsSectionOffset` and disassemble similar to how we do on linux/x64. But not sure if I can avoid disassembling completely. 

<img width="367" alt="image" src="https://github.com/dotnet/runtime/assets/12488060/fb02496b-bbc5-405a-8ef7-8fdd36cdec36">

The final code pattern for osx/x64 is the following (similar to osx/arm64)

<img width="384" alt="image" src="https://github.com/dotnet/runtime/assets/12488060/9c66b0fa-bea9-42f3-88d1-98f1ec42e0e5">

Reply:
I meant `lea rax, _t_ThreadStatics@TLVP[rip]`. You can then just call the helper to get the address without disassembling it, just like you do on arm64. (This assumes that the linker won't choke on `lea rax, _t_ThreadStatics@TLVP[rip]`.)

Reply:
I just realized something. Let me retry this.

Reply:
I was not correctly having the instructions, but yes, we end up with the linker error:

```
ld: TLV load reloc does not point to a movq instruction in '_GetThreadVarsSectionOffsetNew' from ../../../vm/wks/CMakeFiles/cee_wks_core.dir/__/amd64/asmhelpers.S.o for architecture x86_64
  clang: error: linker command failed with exit code 1 (use -v to see invocation)
```